File: pythonproject/ui/grafico_linha.py
from PyQt6.QtWidgets import QWidget
from PyQt6.QtGui import QPainter, QColor, QFont
from PyQt6.QtCore import Qt
from typing import List

class GraficoLinha(QWidget):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.notas: List[float] = []
        self.PADDING = 40
        self.setMinimumSize(400, 300)
        
    def set_notas(self, notas: List[float]):
        self.notas = notas.copy() if notas else []
        if self.notas:
            self.notas.sort()
        self.update()
        
    def paintEvent(self, event):
        painter = QPainter(self)
        painter.setRenderHint(QPainter.RenderHint.Antialiasing)
        
        if not self.notas or len(self.notas) < 2:
            painter.setFont(QFont("Arial", 10))
            painter.setPen(QColor(128, 128, 128))
            painter.drawText(20, 30, "Selecione um curso para exibir o gráfico de notas.")
            return

        nota_minima = self.notas[0]
        nota_maxima = self.notas[-1]
        range_notas = (nota_maxima - nota_minima) if (nota_maxima - nota_minima) > 0 else 1
        
        largura_desenhavel = self.width() - 2 * self.PADDING
        altura_desenhavel = self.height() - 2 * self.PADDING

        painter.setPen(QColor(0, 0, 0))
        painter.drawLine(self.PADDING, self.height() - self.PADDING, self.PADDING, self.PADDING)  # Eixo Y
        painter.drawLine(self.PADDING, self.height() - self.PADDING, self.width() - self.PADDING, self.height() - self.PADDING)  # Eixo X

        painter.setPen(QColor(255, 0, 0))

        
        for i in range(len(self.notas) - 1):
            x1 = self.PADDING + int(i * (largura_desenhavel / (len(self.notas) - 1)))
            y1 = self.height() - self.PADDING - int(((self.notas[i] - nota_minima) / range_notas) * altura_desenhavel)
            x2 = self.PADDING + int((i + 1) * (largura_desenhavel / (len(self.notas) - 1)))
            y2 = self.height() - self.PADDING - int(((self.notas[i + 1] - nota_minima) / range_notas) * altura_desenhavel)
            
            painter.drawLine(x1, y1, x2, y2)

        painter.setPen(QColor(0, 0, 0))
        painter.setFont(QFont("Arial", 8))
        painter.drawText(self.PADDING - 35, self.height() - self.PADDING + 15, f"{nota_minima:.2f}")
        painter.drawText(self.PADDING - 35, self.PADDING, f"{nota_maxima:.2f}")

File: javaproject/src/main/java/ui/GraficoLinha.java
package ui;

import java.awt.BasicStroke;
import java.awt.Color;
import java.awt.Font;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.RenderingHints;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;


public class GraficoLinha extends Grafico {

    private List<Double> notas;
    private final int PADDING = 40;

    public GraficoLinha() {
        super();
        this.notas = new ArrayList<>();
    }
    
    
     //Define os dados e redesenha o componente.
    
    public void setNotas(List<Double> notas) {
        this.notas = (notas != null) ? new ArrayList<>(notas) : new ArrayList<>();
        if (!this.notas.isEmpty()) {
            Collections.sort(this.notas);
        }
        repaint(); 
    }

    @Override
    protected void paintComponent(Graphics g) {
        super.paintComponent(g);
        Graphics2D g2d = (Graphics2D) g;
        g2d.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);

        if (notas == null || notas.size() < 2) {
            g2d.setFont(new Font("Arial", Font.PLAIN, 14));
            g2d.setColor(Color.GRAY);
            g2d.drawString("Selecione um curso para exibir o gráfico de notas.", 20, 30);
            return;
        }

        // ---------------------------------------------
        double notaMinima = notas.get(0);
        double notaMaxima = notas.get(notas.size() - 1);
        double rangeDeNotas = (notaMaxima - notaMinima) == 0 ? 1 : (notaMaxima - notaMinima);
        int larguraDesenhavel = getWidth() - 2 * PADDING;
        int alturaDesenhavel = getHeight() - 2 * PADDING;

        // ---------------------------------------------
        g2d.setColor(Color.BLACK);
        g2d.drawLine(PADDING, getHeight() - PADDING, PADDING, PADDING); // Eixo Y
        g2d.drawLine(PADDING, getHeight() - PADDING, getWidth() - PADDING, getHeight() - PADDING); // Eixo X

        // ---------------------------------------------
        g2d.setStroke(new BasicStroke(2f));
        g2d.setColor(Color.BLACK); // A linha será preta

        for (int i = 0; i < notas.size() - 1; i++) {// Calcula as coordenadas
            int x1 = PADDING + (int) (i * (double) larguraDesenhavel / (notas.size() - 1));
            int y1 = getHeight() - PADDING - (int) (((notas.get(i) - notaMinima) / rangeDeNotas) * alturaDesenhavel);
            int x2 = PADDING + (int) ((i + 1) * (double) larguraDesenhavel / (notas.size() - 1));
            int y2 = getHeight() - PADDING - (int) (((notas.get(i + 1) - notaMinima) / rangeDeNotas) * alturaDesenhavel);
            
            // Desenha a linha de conexão
            g2d.drawLine(x1, y1, x2, y2);
        }

        // Desenha Nota min e max
        g2d.setColor(Color.BLACK);
        g2d.setFont(new Font("Arial", Font.PLAIN, 12));
        g2d.drawString(String.format("%.2f", notaMinima), PADDING - 35, getHeight() - PADDING + 5);
        g2d.drawString(String.format("%.2f", notaMaxima), PADDING - 35, PADDING);
    }
}